## Applied Linear Algebra Home Assignment 2

### 1. Find the best approximation matrix $A_1$ of rank 2 of the matrix A in the norm $||\cdot||_2$ and find $||A-A_1||_2$, where  
$$A = \begin{bmatrix}-81 & -30 & 60 & -48\\
78 & -42 & 24 & 36\\
-42 & 21 & -72 & -12
\end{bmatrix}$$

### Solution:  

Let us find at first SVD decomposition of this matrix $A$, which equals
$$ A = U\Sigma V^* $$
After we find matrix $\Sigma$ we should reduce its rank by 1 (to get the rank 2 and $\Sigma_r$) and then get the matrix $A_1$, which equals
$$ A_1 = R\Sigma_r Q^*, $$
where r = 2.

In [5]:
# Finding SVD of matrix A
import numpy as np
import sympy as sp

A = np.array([[-81, -30, 60, -48], 
              [78, -42, 24, 36],
              [-42, 21, -72, -12]])
U, s, Vh = np.linalg.svd(A, full_matrices=False)
print('Matrix U:\n{}\nMatrix Sigma:\n{}\nMatrix Vh:\n{}'.format(U,np.diag(s),Vh))


Matrix U:
[[ 0.66666667  0.66666667 -0.33333333]
 [-0.66666667  0.33333333 -0.66666667]
 [ 0.33333333 -0.66666667 -0.66666667]]
Matrix Sigma:
[[135.   0.   0.]
 [  0. 108.   0.]
 [  0.   0.  27.]]
Matrix Vh:
[[-8.88888889e-01  1.11111111e-01  0.00000000e+00 -4.44444444e-01]
 [-6.66133815e-16 -4.44444444e-01  8.88888889e-01 -1.11111111e-01]
 [ 1.11111111e-01  8.88888889e-01  4.44444444e-01 -2.15105711e-16]]


$$ A = U \Sigma V^* = \begin{bmatrix} 2/3 & 2/3 & -1/3 \\ -2/3 & 1/3 & -2/3 \\ 1/3 & -2/3 & -2/3 \end{bmatrix} \cdot \begin{bmatrix} 135 & 0 & 0 \\ 0 & 108 & 0 \\ 0 & 0 & 27 \end{bmatrix} \cdot \begin{bmatrix} -8/9 & 1/9 & 0 & -4/9 \\ 0 & -4/9 & 8/9 & -1/9 \\1/9 & 8/9 & 4/9 & 0\end{bmatrix}$$


In [6]:
A @ A.T

array([[13365, -5346,  -972],
       [-5346,  9720, -6318],
       [ -972, -6318,  7533]])

Now, let us reduce rank of matrix $\Sigma$.

In [2]:
# Reducing rank of matrix Sigma to rnk=2
S_rank = np.diag(s)
S_rank[2,2] = 0
S_rank

array([[135.,   0.,   0.],
       [  0., 108.,   0.],
       [  0.,   0.,   0.]])

$$ \Sigma_{rank} = \begin{bmatrix} 135 & 0 & 0 \\ 0 & 108 & 0 \\ 0 & 0 & 0 \end{bmatrix}$$
where $rank=2$

Now, let us compute matrix $A_r$ by using matrix $\Sigma_r$.

In [3]:
# computing matrix A_1
A_r = U @ S_rank @ Vh

# checking matrix and its rank
print('Matrix A_r:\n{}\nIts rank: {}'.format(A_r,np.linalg.matrix_rank(A_r)))

Matrix A_r:
[[-80. -22.  64. -48.]
 [ 80. -26.  32.  36.]
 [-40.  37. -64. -12.]]
Its rank: 2


$$A_r = \begin{bmatrix} -80 & -22 & 64 & -48\\ 80 & -26 & 32 & 36\\ -40 & 37 & -64 & -12\end{bmatrix}$$

So we have matrices $A$ and $A_r$. Let us find  $||A-A_r||_2$.

In [4]:
A_r = A_r.astype(int)
A = A.astype(int)
diff = A - A_r
print("Difference of two matrices:\n{}".format(diff))
_,s_diff,_ = np.linalg.svd(diff)
print('Second norm: {}'.format(max(s_diff)))

Difference of two matrices:
[[ -1  -8  -3   0]
 [ -2 -16  -7   0]
 [ -3 -16  -9   0]]
Second norm: 26.9491099563613


$$||A-A_r||_2 = ||\Sigma - \Sigma_r||_2 = \left\lVert\begin{bmatrix}
-1 & -8 & -3 & 0\\-2 & -16 & -7 & 0\\-3 & -16 & -9 & 0\end{bmatrix}\right\rVert_2 = max_{1\leq i\leq m}{\ \sigma_i}=26.95$$

### Answer:
$||A-A_r||_2\approx\sigma_{r+1}=27$

### 2. Estimate the relative error of the approximate solution $(1, 1)$ of the system $Ax = b$ in the norms $|\cdot|_1$ and $|\cdot|_2$ using the condition number of the matrix $A$, where
$$A = \begin{pmatrix} -2.99 & 0.05 \\ 1.04 & -6.18\end{pmatrix}, b = \begin{pmatrix} -3.03 \\-5.09\end{pmatrix}$$

### Solution:
$$\hat{A} = \begin{pmatrix} -3 & 0 \\ 1 & -6\end{pmatrix}, \hat{b} = \begin{pmatrix} -3 \\-5\end{pmatrix}, \hat{x} = \begin{pmatrix} 1 \\1\end{pmatrix}$$

$$\delta x \leq \varkappa (A) (\delta b + \delta A)$$

Condition number:  
$$\varkappa(A) \approx \varkappa(\hat{A}) \approx ||\hat{A}||\cdot||\hat{A}^{-1}||$$

$$\delta A = \dfrac{||\Delta A||}{||\hat{A}||}, \ \delta b = \dfrac{||\Delta b||}{||\hat{b}||}

So let us find at first the inverse matrix of $\hat{A}$.

In [5]:
A = np.array([[-3,0],[1,-6]])
b = np.array([[-3,-5]])
x = np.array([[1,1]])
A_inv = np.linalg.inv(A)
A_inv

array([[-0.33333333, -0.        ],
       [-0.05555556, -0.16666667]])

$$A^{-1} = \begin{pmatrix} -1/3 & 0\\ -1/18 & -1/6\end{pmatrix}$$

In [6]:
kappa_1 = max(np.sum(np.abs(A), axis=0))*max(np.sum(np.abs(A_inv), axis=0))
print('kappa 1 of matrix A: ', kappa_1)

kappa 1 of matrix A:  2.333333333333333


For the norm $|\cdot|_1$ we have a condition number:
$$\varkappa_1 = max\{4,6\} \cdot max\{7/18, 3/18\}= 6 \cdot \dfrac{7}{18} = \dfrac{21}{9}$$

For the norm $|\cdot|_2$ we have a condition number:
$$\varkappa_2 = \frac{|\lambda_{max}(A^*A)|}{|\lambda_{min}(A^*A)|}$$
Let us find $A^*A$ and then $\varkappa_2$.

In [7]:
AA = A.T @ A
eigvals = sorted(np.linalg.eigvals(AA))
kappa_2 = max(eigvals) / min(eigvals)
print(f"Eigen values: {eigvals}")
print('Kappa 2 of matrix A: ', kappa_2)

Eigen values: [8.682178936723648, 37.317821063276355]
Kappa 2 of matrix A:  4.298209163304667


In [8]:
_,singval,_ = np.linalg.svd(AA, full_matrices=True)
print('Singular Values: {}'.format(singval))
kappa_2 = np.sqrt(max(singval)/min(singval))
print('Kappa 2 of matrix A: ', kappa_2)

Singular Values: [37.31782106  8.68217894]
Kappa 2 of matrix A:  2.073212281293131


So, we have eigenvalues:  
$ \lambda_{max} = 37.318,\, \lambda_{min} = 8.68$ and $\varkappa_2 = 4.3$

Let us now find $\Delta b, \delta_1 b,\delta_2 b$ and $\Delta A, \delta_1 A,\delta_2 A$.

In [9]:
A_real = np.array([[-2.99, 0.05],[1.04, -6.18]])
b_real = np.array([[-3.03,-5.09]])
Delta_b = b - b_real
Delta_A = A - A_real


delta_1_b = np.sum(np.abs(Delta_b))/np.sum(np.abs(b))
delta_2_b = np.sqrt(np.sum(Delta_b**2))/np.sqrt(np.sum(b**2))

print("""delta_1_b: {}\n
delta_2_b: {}\n""".format(delta_1_b,delta_2_b))

delta_1_b: 0.014999999999999958

delta_2_b: 0.01626978433639918



In [10]:
delta_1_A = max(np.sum(np.abs(Delta_A),axis=0))/max(np.sum(np.abs(A),axis=0))

dAdA = Delta_A.T @ Delta_A
_,singval_dAdA,_ = np.linalg.svd(dAdA)
delta_2_A = np.sqrt(max(singval_dAdA)/max(singval))
delta_2_A
print("""delta_1_A: {}\n
delta_2_A: {}\n""".format(delta_1_A,delta_2_A))

delta_1_A: 0.03833333333333328

delta_2_A: 0.031146040806455853



$$\Delta b = \begin{pmatrix}0.03 \\ 0.09\end{pmatrix}$$

$$\Delta A = \begin{pmatrix}0.01 & 0.05\\0.04 & -0.18\end{pmatrix}$$

$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = \frac{0.12}{8}= 0.015$$

$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = \frac{\sqrt{0.03^2+0.09^2}}{\sqrt{3^2+5^2}}= 0.0163$$

$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = \frac{0.23}{6} = 0.0383$$

For $\delta_2 A$ we find $\Delta A^* \Delta A$ and then:  
$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = \frac{\sqrt{0.0362}}{\sqrt{37.32}}= 0.0311$$

Let's check the relative errors of solutions in our norms.

In [11]:
norm1_upper_bound = kappa_1*(delta_1_b+delta_1_A)
norm2_upper_bound = kappa_2*(delta_2_b+delta_2_A)

norm1_lower_bound = 1/kappa_1*(delta_1_b+delta_1_A)
norm2_lower_bound = 1/kappa_2*(delta_2_b+delta_2_A)

print("""Error 1 upper bound: {}\n
Error 1 lower bound: {}\n
Error 2 upper bound: {}\n
Error 2 lower bound: {}\n""".format(norm1_upper_bound,
                                   norm1_lower_bound,
                                   norm2_upper_bound,
                                   norm2_lower_bound))

Error 1 upper bound: 0.1244444444444442

Error 1 lower bound: 0.02285714285714282

Error 2 upper bound: 0.09830307101381466

Error 2 lower bound: 0.0228707043512593



### Answer:
Norm $|\cdot|_2$ gives the lowest relative error $\delta_2 x \in [0.02287, 0.0983]$

Norm $|\cdot|_1$ gives the lowest relative error $\delta_1 x \in [0.02285, 0.124]$


### 3. Solve the system approximately and estimate the error of the solution in the norms $|\cdot|_1$, $|\cdot|_2$ and $|\cdot|_{\infty}$:
$$ \begin{cases} -2(6+\varepsilon_1)x+2(1+\varepsilon_2)y = 4+\varepsilon_3 \\ 4x + (4+\varepsilon_1)y = -3 + \varepsilon_4 \end{cases}$$

### where the unknown numbers $\varepsilon_j$ satisfy the conditions $|\varepsilon_j| < 0.05$ for all $j=1,...,4$.

### Solution:
$$\hat{A} = \begin{pmatrix} -12 & 2 \\ 4 & 4\end{pmatrix}, \hat{b} = \begin{pmatrix} 4 \\-3\end{pmatrix}, \hat{x} = \begin{pmatrix} -\dfrac{11}{28} \\[0.5cm] -\dfrac{5}{14}\end{pmatrix}$$
Condition number:  
$$\varkappa(A) \approx \varkappa(\hat{A}) \approx ||\hat{A}||\cdot||\hat{A}^{-1}||
$$

So let us find at first the inverse matrix of $\hat{A}$.

In [75]:
A = np.array([[-12, 2],[4, 4]])
b = np.array([[4, -3]])
x = np.array([[-11/28, -5/14]])
A_inv = np.linalg.inv(A)
A_inv

array([[-0.07142857,  0.03571429],
       [ 0.07142857,  0.21428571]])

$$A^{-1} = \begin{pmatrix}-\dfrac{1}{14} & \dfrac{1}{28}\\[0.5cm] \dfrac{1}{14} & \dfrac{3}{14}\end{pmatrix}$$

In [82]:
kappa_1 = max(np.sum(np.abs(A),axis=0))*max(np.sum(np.abs(A_inv),axis=0))
print('Kappa 1 of matrix A: ', kappa_1)
kappa_inf = max(np.sum(np.abs(A),axis=1))*max(np.sum(np.abs(A_inv),axis=1))
print('Kappa inf of matrix A: ', kappa_inf)

Kappa 1 of matrix A:  4.0
Kappa inf of matrix A:  4.0


For the norm $|\cdot|_1$ we have a condition number:
$$\varkappa_1 = max\{16, 8\} \cdot max\{\dfrac{4}{28}, \frac{7}{28}\}= 4$$

For the norm $|\cdot|_{\infty}$ we have a condition number:
$$\varkappa_{\infty} = max\{14,8\} \cdot max\{\dfrac{8}{28}, \frac{3}{28}\}=4 $$

For the norm $|\cdot|_2$ we have a condition number:
$$\varkappa_2 = \sqrt{\frac{\lambda_{max}(A^*A)}{\lambda_{min}(A^*A}}$$
Let us find $A^*A$ and then $\varkappa_2$.

In [83]:
AA = A.T @ A
_,singval,_ = np.linalg.svd(AA, full_matrices=True)
print('Singular Values: {}'.format(singval))
kappa_2 = np.sqrt(max(singval)/min(singval))
print('Kappa 2 of matrix A: ', kappa_2)

Singular Values: [160.45565982  19.54434018]
Kappa 2 of matrix A:  2.073212281293131


So, we have eigenvalues:  
$ \lambda_{max} = 160.455,\, \lambda_{min} = 19.544$ and $\varkappa_2 = 2.073$

Let us now find $\Delta b$ and $\Delta A$.

In [84]:
e1, e2, e3, e4 = sp.symbols('varepsilon1 varepsilon2 varepsilon3 varepsilon4')
A_real = sp.Matrix([[-2*(6+e1), 2*(1+e2)],[4, 4+e1]])
b_real = sp.Matrix([[4+e3, -3+e4]])

In [85]:
Delta_A = A_real - sp.Matrix(A)
Delta_b = b_real - sp.Matrix(b)
print('Delta A equals:')
Delta_A

Delta A equals:


Matrix([
[-2*varepsilon1, 2*varepsilon2],
[             0,   varepsilon1]])

In [86]:
print('Delta b equals:')
Delta_b
# Delta_b.subs([(e3,0.05),(e4,0.05)])

Delta b equals:


Matrix([[varepsilon3, varepsilon4]])

Because $\varepsilon_j s$ are unknown and lie in interval of $(-0.05, 0.05)$ we will evaluate the solution by the maximum interval. Let us find epsilons, where sum of deltas $A$ and $b$ in terms of norm $|\cdot|_1$ are maximum.

In [87]:
from operator import itemgetter

delta_1_Ab_list = []
for eps1 in np.arange(-0.05,0.05, 0.01).round(2):
    for eps2 in np.arange(-0.05,0.05, 0.01).round(2):
        for eps3 in np.arange(-0.05,0.05, 0.01).round(2):
            for eps4 in np.arange(-0.05,0.05, 0.01).round(2):
                Delta_b_new = Delta_b.subs([(e3,eps3),(e4,eps4)])
                Delta_b_new = np.array(Delta_b_new).astype(np.float64)
                Delta_A_new = Delta_A.subs([(e1, eps1),(e2,eps2)])
                Delta_A_new = np.array(Delta_A_new).astype(np.float64)
                
                delta_1_b = np.sum(np.abs(Delta_b_new))/np.sum(np.abs(b))
                delta_1_A = max(np.sum(np.absolute(Delta_A_new),axis=0))/max(np.sum(np.absolute(A),axis=0))
                delta_1_Ab = delta_1_A + delta_1_b
                
                delta_1_Ab_list.append((delta_1_Ab,eps1,eps2,eps3,eps4))

print(max(delta_1_Ab_list,key=itemgetter(1)))

(0.023035714285714288, 0.04, -0.05, -0.05, -0.05)


So, as we can see
$$\varepsilon_1 = 0.04$$
$$\varepsilon_2 = -0.05$$
$$\varepsilon_3 = -0.05$$
$$\varepsilon_4 = -0.05$$
Let us substitute these epsilons into $A$ and $b$.
We got:

In [92]:
Delta_A = Delta_A.subs([(e1,0.04),(e2,-0.05)])
Delta_A

Matrix([
[-0.08, -0.1],
[    0, 0.04]])

In [89]:
Delta_b = Delta_b.subs([(e3,-0.05),(e4,-0.05)])
Delta_b

Matrix([[-0.05, -0.05]])

For $|\cdot|_1$:

In [95]:
delta_1_b = np.sum(np.abs(Delta_b))/np.sum(np.abs(b))
delta_1_A = max(np.sum(np.abs(Delta_A),axis=0))/max(np.sum(np.abs(A),axis=0))
print('delta_1_A: ', delta_1_A)
print('delta_1_b: ', delta_1_b)

delta_1_A:  0.00875000000000000
delta_1_b:  0.0142857142857143


$$\delta_1 A =\frac{|\Delta A|_1}{|A|_1}\approx \frac{|\Delta A|_1}{|\hat{A}|_1} = \frac{0.14}{16}= 0.00875$$
$$\delta_1 b =\frac{|\Delta b|_1}{|b|_1}\approx \frac{|\Delta b|_1}{|\hat{b}|_1} = \frac{0.1}{7}= 0.0142857$$

For $|\cdot|_2$:

In [96]:
Delta_b_new = np.array(Delta_b).astype(np.float64)
delta_2_b = np.sqrt(np.sum(Delta_b_new**2))/np.sqrt(np.sum(b**2))
Delta_A_new = np.array(Delta_A).astype(np.float64)

dAdA = Delta_A_new.T @ Delta_A_new
_,singval_dAdA,_ = np.linalg.svd(dAdA)
delta_2_A = np.sqrt(max(singval_dAdA)/max(singval))
print('delta_2_A: ', delta_2_A)
print('delta_2_b: ', delta_2_b)

delta_2_A:  0.010417068406560195
delta_2_b:  0.014142135623730952


For $\delta_2 A$ we find $\Delta A^* \Delta A$ and then:  

$$\delta_2 A =\frac{|\Delta A|_2}{|A|_2}\approx \frac{|\Delta A|_2}{|\hat{A}|_2} =\frac{\sigma_1 (\Delta A)}{\sigma_1 (A)} = \frac{\sqrt{0.0174119}}{\sqrt{160.455}}= 0.01042$$
$$\delta_2 b =\frac{|\Delta b|_2}{|b|_2}\approx \frac{|\Delta b|_2}{|\hat{b}|_2} = \frac{\sqrt{0.05^2+0.05^2}}{\sqrt{3^2+4^2}}= 0.014142$$

For $|\cdot|_{\infty}$:

In [108]:
delta_inf_b = np.max(np.abs(Delta_b))/np.max(np.abs(b))
delta_inf_A = max(np.sum(np.absolute(Delta_A),axis=1))/max(np.sum(np.absolute(A),axis=1))
print('delta_inf_A: ', delta_inf_A)
print('delta_inf_b: ', delta_inf_b)

delta_inf_A:  0.0128571428571429
delta_inf_b:  0.0125000000000000


$$\delta_{\infty} A =\frac{|\Delta A|_{\infty}}{|A|_{\infty}}\approx \frac{|\Delta A|_{\infty}}{|\hat{A}|_{\infty}} = \frac{0.18}{14}= 0.01286$$
$$\delta_{\infty} b =\frac{|\Delta b|_{\infty}}{|b|_{\infty}}\approx \frac{|\Delta b|_{\infty}}{|\hat{b}|_{\infty}} = \frac{0.05}{4}= 0.0125$$

Let's check the relative errors of solutions in our norms.

In [114]:
norm1_upper_bound = kappa_1*(delta_1_b+delta_1_A)
norm2_upper_bound = kappa_2*(delta_2_b+delta_2_A)
norm_inf_upper_bound = kappa_inf*(delta_inf_b+delta_inf_A)

print("""Error 1 upper bound: {}\n
Error 2 upper bound: {}\n
Error Inf upper bound: {}""".format(norm1_upper_bound,
                                   norm2_upper_bound,
                                   norm_inf_upper_bound))

Error 1 upper bound: 0.0921428571428572

Error 2 upper bound: 0.07036898727173792

Error Inf upper bound: 0.101428571428571


### Answer:
Approximate solution:
$$\hat{x} = \begin{pmatrix} -\dfrac{11}{28} \\[0.5cm] -\dfrac{5}{14}\end{pmatrix}$$
Relative errors:
$$\delta_{1} x \leq  0.092$$
$$\delta_{2} x \leq 0.07036$$
$$\delta_{\infty} x \leq 0.10142$$

### 4. Find the approximate inverse matrix to the matrix $A$ and evaluate the approximation error with respect to the uniform norm $||\cdot||_1$, if the elements of the matrix $A$ are known with an absolute error of 0.01:
$$A \approx \begin{pmatrix}-4 & -7\\-5 & 9\end{pmatrix}$$

### Solution:
$$A \approx \begin{pmatrix}-4 & -7\\-5 & 9\end{pmatrix}$$
$$\hat{A} = \begin{pmatrix}-4 & -7\\-5 & 9\end{pmatrix}, \varepsilon = 0.01$$

The real matrix $A$ looks like
$$A = \begin{pmatrix}-4 & -7\\-5 & 9\end{pmatrix}+\begin{pmatrix}\varepsilon_{11} & \varepsilon_{12}\\\varepsilon_{21} & \varepsilon_{22}\end{pmatrix},$$
where $|\varepsilon_{ij}|\leq 0.01$

Let us find inverse matrix of A.
$$\hat{A}^{-1} = \frac{1}{det \hat A}\begin{pmatrix} 9 & 7\\5 & -4\end{pmatrix}=\frac{1}{-71}\begin{pmatrix} 9 & 7\\5 & -4\end{pmatrix}$$

In [117]:
A_4 = np.array([[-4, -7],[-5, 9]])
A_4_inv = np.linalg.inv(A_4)
A_4_inv

array([[-0.12676056, -0.09859155],
       [-0.07042254,  0.05633803]])

Let us find next the conditional number $\varkappa$ in norm $|\cdot|_1$.  
And let us assume that $\varkappa (A) \approx \varkappa(\hat A)$

$$\varkappa_1(A) \approx \varkappa(\hat{A}) \approx ||\hat{A}||_1\cdot||\hat A^{-1}||_1=max\{16,9\}\cdot max\{0.197, 0.155\} = 16 \cdot 0.197 = 3.155
$$

So,
$$ \delta A^{-1} \leq \frac{\varkappa(\hat A) \delta \varepsilon}{1-\varkappa (\hat A)\delta \epsilon}, \delta \varepsilon = \frac {||\varepsilon||}{||\hat A||}$$

In our case with norm $|\cdot|_1$ we have:  
$$\delta \varepsilon = \frac {||\varepsilon||_1}{||\hat A||_1}=\frac {max\{0.2,0.2\}}{16}=0.0125$$
Approximation error:
$$ \delta A^{-1} \leq \frac{\varkappa_1(\hat A) \delta \varepsilon}{1-\varkappa_1 (\hat A)\delta \varepsilon}=\frac{3.155\cdot 0.0125}{1-3.155\cdot 0.0125}=0.0411$$

### Answer:
$$A^{-1} \approx \frac{1}{-71}\begin{pmatrix} 9 & 7\\5 & -4\end{pmatrix}$$
$$ \delta A^{-1} \leq 0.0411$$

### 5. Use simple iteration method for finding the solution of the given linear system.
$$ 
\left\{\begin{array}{c} 26x + 2y + 6z = 2, \\ 2x + 19y + 8z = 4, \\ 3x + 4y + 20z=3. \end{array}\right.
$$

### Determine the iteration number after which the approximation error for each coordinate does not exceed $0.01$ and find the corresponding approximate solution. Start with $x_0 = [0, 0, 0]^T$.

### Solution:

So we have matrix A and vector b.
$$A = \begin{pmatrix}26 & 2 & 6\\2 & 19 & 8\\ 3 & 4 & 20\end{pmatrix}$$
$$B = \begin{pmatrix}2\\4\\3\end{pmatrix}$$

Let us bring our system of linear equations to the next form:
$$ x_{k+1} = P x_k+\bar b, $$
where $P=E-CA$ and $\bar b = CB$.  
We need to choose such $C$ that our iteration process converges, i.e. $||P||<1$.  
All calculations will be in terms of norm $|\cdot|_{\infty}$.

In [124]:
A = np.array([[26, 2, 6],[2, 19, 8],[3, 4, 20]])

B = np.array([2, 4, 3])

epsilon = 0.01 

Let us find this matrix $C$, matrix $P$ and vector $b$.  
$C = \begin{pmatrix}\frac{1}{26} & 0 & 0\\0 & \frac{1}{19} & 0\\ 0 & 0 & \frac{1}{20}\end{pmatrix}$  
$P = I-CA = \begin{pmatrix} 1 & 0 & 0\\0 & 1 & 0\\ 0 & 0 & 1\end{pmatrix}-\begin{pmatrix}\frac{1}{26} & 0 & 0\\0 & \frac{1}{19} & 0\\ 0 & 0 & \frac{1}{20}\end{pmatrix}\cdot\begin{pmatrix}26 & 2 & 6\\2 & 19 & 8\\ 3 & 4 & 20\end{pmatrix}$ 


Its norm will be $||P||_{\infty}=0.53<1$. It means that iteration process converges.  


$b = CB = \begin{pmatrix}\frac{1}{26} & 0 & 0\\0 & \frac{1}{19} & 0\\ 0 & 0 & \frac{1}{20}\end{pmatrix} \cdot \begin{pmatrix}2\\4\\3\end{pmatrix}= \begin{pmatrix}0.07692308\\0.21052632\\0.15\end{pmatrix}$

In [125]:
C = np.array([[1/26, 0, 0],[0, 1/19, 0],[0, 0, 1/20]])
I = np.eye(3)

P = I - np.matmul(C, A)
b = np.matmul(C,B)
# Norm_infinity of Matrix
P_norm = max(np.sum(np.absolute(P),axis=1)) # 0.5 < 1 -> converges
print('P:\n', P)
print('b:\n', b)
print('Norm_infinity of Matrix P: {}'.format(P_norm))

P:
 [[ 0.         -0.07692308 -0.23076923]
 [-0.10526316  0.         -0.42105263]
 [-0.15       -0.2         0.        ]]
b:
 [0.07692308 0.21052632 0.15      ]
Norm_infinity of Matrix P: 0.5263157894736842


In purpose to find the number of iterations of the process we will use priori condition:
$$ m \geq \frac{ln(\frac{\varepsilon(1-||P||_{\infty})}
{||x^{(1)}-x^{(0)}||_{\infty}})}{ln ||P||_{\infty}}$$  
$$||x^{(1)}-x^{(0)}||_{\infty}=||b||_{\infty} = ||\begin{bmatrix} 0.07692308 & 0.21052632 & 0.15\end{bmatrix}^T||_{\infty}=0.21052632$$  
$$ m \geq \frac{ln(\frac{0.01*0.47)}{0.21052632})}{ln 0.5}=5.49$$
Let us do the iteration process with number of iterations of 6 using this formula:
$$ x_{k+1} = P x_k+\bar b$$

In [132]:
# Initial and next values of x 
x_0 = np.array([0,0,0])
x_1 = np.matmul(P, x_0) + b

# Norm_infinity of Vector
norm_diff = np.max(np.abs(x_1 - x_0))
#apriori condition
m = int(np.log(epsilon*(1-P_norm)/norm_diff)/np.log(P_norm))
print(m)
print('Number of iterations: {:.0f}'.format(m+1))

#loop with number of iterations = 6
x = x_0
for i in range(int(m)+1):
    x = np.matmul(P, x) + b
    print(x)
#     print('$$x_{}={}^T$$'.format(i+1, sp.latex(sp.Matrix(x.round(4)).T)))
    
print('\nSolution: {}'.format(x))

5
Number of iterations: 6
[0.07692308 0.21052632 0.15      ]
[0.02611336 0.13927126 0.09635628]
[0.04397384 0.16720648 0.11822874]
[0.03677748 0.15611697 0.10996263]
[0.03953809 0.16035495 0.11325998]
[0.03845116 0.158676   0.1119983 ]

Solution: [0.03845116 0.158676   0.1119983 ]


### Answer:
$$x_{ans}=\left[\begin{matrix}0.03845116 & 0.158676 & 0.1119983\end{matrix}\right]^T$$

### 6. Find the most influential vertex in the graph:
$$A = \begin{bmatrix}1 & 0 & 1 & 0 & 0\\0 & 1 & 0 & 0 & 1\\0 & 1 & 0 & 1 & 0\\1 & 1 & 0 & 0 & 0\\0 & 1 & 0 & 0 & 1\end{bmatrix}$$
### using the PageRank algorithm with $β = 0.15$, where the graph adjacency matrix is defined as follows

In [133]:
A = np.array([[1, 0, 1, 0, 0],
              [0, 1, 0, 0, 1],
              [0, 1, 0, 1, 0],
              [1, 1, 0, 0, 0],
              [0, 1, 0, 0, 1]])

Let us build matrix of possibilities $P$.

In [134]:
P = np.array([A[0]/2, A[1]/2, A[2]/2, A[3]/2, A[4]/2])
sp.Matrix(P.round(3))

Matrix([
[0.5,   0, 0.5,   0,   0],
[  0, 0.5,   0,   0, 0.5],
[  0, 0.5,   0, 0.5,   0],
[0.5, 0.5,   0,   0,   0],
[  0, 0.5,   0,   0, 0.5]])

Now, let us build matrix $Q$ in order to find matrix $\hat P$.  

$$\hat P = P\cdot(1-\beta)+Q\cdot\beta,$$
where $\beta = 0.15$

In [135]:
n = 5

Q = np.ones((5,5))/n

beta = 0.15

P_hat = P * (1 - beta) + Q * beta

sp.Matrix(P_hat)

Matrix([
[0.455,  0.03, 0.455,  0.03,  0.03],
[ 0.03, 0.455,  0.03,  0.03, 0.455],
[ 0.03, 0.455,  0.03, 0.455,  0.03],
[0.455, 0.455,  0.03,  0.03,  0.03],
[ 0.03, 0.455,  0.03,  0.03, 0.455]])

We will begin with $X_0$ with equal possibilities.
$$X_0 = \left[\begin{matrix}0.2\\0.2\\0.2\\0.2\\0.2\end{matrix}\right]$$
Iteration will process until the difference between $x_{i+1}$ and $x_{i}$ will be small enough ($\varepsilon=0.001$).

In [136]:
x = np.ones(5)/n
cond = True
i = 1
while cond:
    x_new = np.matmul(P_hat.T, x)
    print(x_new.round(4))
#     print('$$x_{}={}^T$$'.format(i,sp.latex(sp.Matrix(x_new.round(4)).T)))
    diff = x_new - x
    if (diff < 0.001).all():
        cond = False
    x = x_new
    i += 1

[0.2   0.37  0.115 0.115 0.2  ]
[0.1639 0.37   0.115  0.0789 0.2722]
[0.1332 0.3854 0.0996 0.0789 0.303 ]
[0.1201 0.3984 0.0866 0.0723 0.3225]
[0.1118 0.4039 0.0811 0.0668 0.3364]
[0.1059 0.4075 0.0775 0.0644 0.3446]
[0.1024 0.41   0.075  0.0629 0.3497]
[0.1003 0.4115 0.0735 0.0619 0.3528]
[0.0989 0.4124 0.0726 0.0612 0.3548]
[0.0981 0.413  0.072  0.0609 0.3561]
[0.0975 0.4133 0.0717 0.0606 0.3568]


### Answer:
So the most infuential vertex in this Graph is vertex $2$ with possibility $P=0.4133$.